src/main/java/org/spongepowered/common/entity/SpongeEntityArchetypeBuilder.java

```diff
@@ -0,0 +1,177 @@
+            );
+        } else {
+            builder.entityData(new MemoryDataContainer());
+        }
+        return Optional.of(builder.build());
+    }
+
+    @Override
+    public EntityArchetype.Builder type(EntityType type) {
+        checkNotNull(type, "EntityType cannot be null!");
+        checkArgument(type != UNKNOWN, "EntityType cannot be set to UNKNOWN!");
+        if (this.entityType != type) {
+            this.entityData = new MemoryDataContainer();
+        }
+        this.entityType = type;
+        return this;
+    }
+
+    @Override
+    public EntityArchetype.Builder from(Entity entity) {
+        checkNotNull(entity, "Cannot build an EntityArchetype for a null entity!");
+        this.entityType = checkNotNull(entity.getType(), "Entity is returning a null EntityType!");
+        final net.minecraft.entity.Entity minecraftEntity = EntityUtil.toNative(entity);
+        final NBTTagCompound compound = new NBTTagCompound();
+        minecraftEntity.writeToNBT(compound);
+        compound.removeTag(NbtDataUtil.UUID);
+        this.entityData = NbtTranslator.getInstance().translate(compound);
+        return this;
+    }
+
+    @Override
+    public EntityArchetype.Builder entityData(DataView view) {
+        checkNotNull(view, "Provided DataView cannot be null!");
+        final DataContainer copy = view.copy();
+        SpongeDataManager.getInstance().getValidators(Validations.ENTITY).validate(copy);
+        this.entityData = copy;
+        return this;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public EntityArchetype.Builder setData(DataManipulator<?, ?> manipulator) {
+        if (this.entityData == null) {
+            this.entityData = new MemoryDataContainer();
+        }
+        SpongeDataManager.getInstance().getRawNbtProcessor(NbtDataTypes.ENTITY, manipulator.getClass())
+                .ifPresent(processor -> processor.storeToView(this.entityData, manipulator));
+        return this;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <E, V extends BaseValue<E>> EntityArchetype.Builder set(V value) {
+        if (this.entityData == null) {
+            this.entityData = new MemoryDataContainer();
+        }
+        SpongeDataManager.getInstance().getRawNbtProcessor(NbtDataTypes.TILE_ENTITY, value.getKey())
+                .ifPresent(processor -> processor.offer(this.entityData, value));
+        return this;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <E, V extends BaseValue<E>> EntityArchetype.Builder set(Key<V> key, E value) {
+        if (this.entityData == null) {
+            this.entityData = new MemoryDataContainer();
+        }
+        SpongeDataManager.getInstance().getRawNbtProcessor(NbtDataTypes.TILE_ENTITY, key)
+                .ifPresent(processor -> processor.offer(this.entityData, value));
+        return this;
+    }
+
+    @Override
+    public EntityArchetype build() {
+        checkState(!this.entityData.isEmpty());
```

Comment:
This should be removed - it's absolutely fine for entity data to be empty. It's only during serialization to native format that it should not be empty (ie adding the `EntityType` field)